<launch>
  <arg name="localization"      default="false"/>
  <arg name="database_path"     default="rtabmap.db"/>
  <arg name="com_port" default="/dev/ttyUSB0"/>
  <arg name="preview" default="false"/>
  <arg name="wired_sync_enabled" default="false"/>

  <!-- Arguments for Kinect driver -->
  <arg name="azure_kinect_driver_required" default="false" />
  <arg name="recording_file" default="" />
  <arg name="recording_loop_enabled" default="true" />
  <arg name="open_loop_controller_enabled" default="false"/>
  
  <!-- rosaria -->
  <include file="$(find gait_training_robot)/launch/includes/robot.launch">
    <arg name="com_port" value="$(arg com_port)"/>
    <arg name="preview" value="$(arg preview)"/>
  </include>

  <!-- Kinect1 + rtabmap_ros -->
  <include file="$(find gait_training_robot)/launch/includes/kinect_rtabmap.launch" unless="$(arg open_loop_controller_enabled)">
    <arg name="localization" value="$(arg localization)" />
    <arg name="database_path" value="$(arg database_path)" />
    <arg name="preview" value="$(arg preview)" />
  </include>
  <node name="rtabmap_dummy" pkg="tf" type="static_transform_publisher"  
    args="0 0 0 0 0 0 map odom 100" if="$(arg open_loop_controller_enabled)"/>

  <!-- distance_controller -->
  <arg name="enable_distance_controller" default="false"/>
  <arg name="v_max" default="1.2"/>

  <!-- goal_generator -->
  <arg name="enable_goal_generator" default="false"/>
  <arg name="max_num_laps" default="3"/>
  <arg name="delay_start_secs" default="0.0"/>
  <arg name="suffix" default="_cw"/>
  <arg name="stop_upon_completion" default="true"/>

  <!-- open_loop_controller -->
  <arg name="open_loop_controller_mode" default="straight"/>
  <arg name="open_loop_controller_terminal_distance" default="5.0"/>
  <arg name="open_loop_controller_wait_for_required_msgs" default="false"/>
  <node pkg="gait_training_robot" type="open_loop_controller.py" name="open_loop_controller" output="screen" 
      if="$(arg open_loop_controller_enabled)" required="true">
    <param name="mode" value="$(arg open_loop_controller_mode)"/>
    <param name="terminal_distance" value="$(arg open_loop_controller_terminal_distance)"/>
    <param name="wait_for_required_msgs" value="$(arg open_loop_controller_wait_for_required_msgs)"/>
    <param name="v_max" value="$(arg v_max)"/>
  </node>

  <!-- Specify the where azure kinect is positioned on the robot.
    This is needed even if the azure kinect driver is included in the calling launch file -->
  <node name="tf_broadcaster_azure_kinect" pkg="tf" type="static_transform_publisher"  
    args="-0.028 0.023 0.0 3.1416 0 0 camera_mount_top camera_base 100" />

  <!-- Record trajectory
  <node name="trajectory_server" pkg="hector_trajectory_server" type="hector_trajectory_server">
    <rosparam file="$(find gait_training_robot)/params/slam/gmapping_params.yaml" command="load" />
  </node>-->

  <!-- move_base -->
  <!-- Unless open_loop_controller_enabled -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" unless="$(arg open_loop_controller_enabled)">
    <param name="clearing_rotation_allowed" value="false" />
    <rosparam file="$(find gait_training_robot)/params/slam/costmap_common_params.yaml" command="load" ns="global_costmap" /> 
    <rosparam file="$(find gait_training_robot)/params/slam/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find gait_training_robot)/params/slam/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find gait_training_robot)/params/slam/global_costmap_params.yaml" command="load" /> 
    <rosparam file="$(find gait_training_robot)/params/slam/base_local_planner_params.yaml" command="load" />
    <remap from="/cmd_vel" to="/distance_controller/cmd_vel_in" if="$(arg enable_distance_controller)" />
  </node>

  <!-- Kinect driver launch file -->
  <!-- Have to delay-launch this because openni will not work -->
  <group unless="$(arg wired_sync_enabled)">
    <include file="$(find timed_roslaunch)/launch/timed_roslaunch.launch"
            if="$(eval arg('enable_distance_controller') or 
                        arg('azure_kinect_driver_required'))">
      <arg name="time" value="2" />
      <arg name="pkg" value="azure_kinect_ros_driver" />
      <arg name="file" value="driver.launch" />
      <arg name="value" 
          value="depth_mode:=WFOV_2X2BINNED
                body_tracking_enabled:=true fps:=30 rgb_point_cloud:=false
                recording_file:=$(arg recording_file)
                recording_loop_enabled:=$(arg recording_loop_enabled)
                required:=true
                output:=screen" />
    </include>
  </group>
  <group if="$(arg wired_sync_enabled)">
    <include file="$(find timed_roslaunch)/launch/timed_roslaunch.launch"
            if="$(eval arg('enable_distance_controller') or 
                        arg('azure_kinect_driver_required'))">
      <arg name="time" value="2" />
      <arg name="pkg" value="azure_kinect_ros_driver" />
      <arg name="file" value="driver.launch" />
      <arg name="value" 
          value="depth_mode:=WFOV_2X2BINNED
                body_tracking_enabled:=true fps:=30 rgb_point_cloud:=false
                wired_sync_mode:=2 subordinate_delay_off_master_usec:=800
                recording_file:=$(arg recording_file)
                recording_loop_enabled:=$(arg recording_loop_enabled)
                required:=true
                output:=screen" />
    </include>
  </group>

  <!-- Unless open_loop_controller_enabled -->
  <group unless="$(arg open_loop_controller_enabled)">
    <!-- If distance controller is enabled: -->
    <group if="$(arg enable_distance_controller)">
      <!-- Distance controller node -->
      <node pkg="gait_training_robot" type="distance_controller" name="distance_controller" output="screen" >
        <remap from="/distance_controller/cmd_vel_out" to="/cmd_vel" />
        <rosparam file="$(find gait_training_robot)/params/distance_controller.yaml" command="load" />
      </node>
    </group>

    <!-- goal_generator -->
    <node pkg="gait_training_robot" type="goal_generator" name="goal_generator" if="$(arg enable_goal_generator)" output="screen" >
      <param name="max_num_laps" value="$(arg max_num_laps)" />
      <param name="delay_start_secs" value="$(arg delay_start_secs)" />
      <param name="suffix" value="$(arg suffix)" />
      <param name="stop_upon_completion" value="$(arg stop_upon_completion)" />
    </node>
  </group>


</launch>